File: fedml/src/main/java/nl/tudelft/trustchain/fedml/ai/dataset/mnist/CustomMnistManager.java
package nl.tudelft.trustchain.fedml.ai.dataset.mnist;


import org.deeplearning4j.datasets.mnist.MnistImageFile;
import org.deeplearning4j.datasets.mnist.MnistLabelFile;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.stream.IntStream;

import nl.tudelft.trustchain.fedml.ai.Behaviors;
import nl.tudelft.trustchain.fedml.ai.dataset.DatasetManager;


public class CustomMnistManager extends DatasetManager {
    private final MnistImageFile mnistImageFile;
    private final MnistLabelFile mnistLabelFile;
    private final byte[][] imagesArr;
    private final int[] labelsArr;

    public CustomMnistManager(String imagesFile, String labelsFile, int numExamples, List<Integer> iteratorDistribution, int maxTestSamples, int seed, Behaviors behavior) throws IOException {
        mnistImageFile = new MnistImageFile(imagesFile, "r");
        mnistLabelFile = new MnistLabelFile(labelsFile, "r");
        final byte[][] tmpImagesArr = loadImages(mnistImageFile, numExamples);
        final int[] tmpLabelsArr = loadLabels(mnistLabelFile, numExamples);
        int[] tmpLabelsArr2 = Arrays.copyOf(tmpLabelsArr, tmpLabelsArr.length);
        if (behavior == Behaviors.LABEL_FLIP) {
            IntStream.range(0, tmpLabelsArr.length).filter(i -> tmpLabelsArr[i] == 1).forEach(i -> tmpLabelsArr2[i] = 2);
            IntStream.range(0, tmpLabelsArr.length).filter(i -> tmpLabelsArr[i] == 2).forEach(i -> tmpLabelsArr2[i] = 1);
        }

        final int totalExamples = calculateTotalExamples(iteratorDistribution, maxTestSamples, tmpLabelsArr2);
        final ImageLabelContainer res = sampleData(tmpImagesArr, tmpLabelsArr2, totalExamples, iteratorDistribution, maxTestSamples, seed);
        imagesArr = res.images;
        labelsArr = res.labels;
    }

    private static int[] loadLabels(MnistLabelFile mnistLabelFile, int numExamples) throws IOException {
        return mnistLabelFile.readLabels(numExamples);
    }

    private static byte[][] loadImages(MnistImageFile mnistImageFile, int numExamples) throws IOException {
        return mnistImageFile.readImagesUnsafe(numExamples);
    }

    private static ImageLabelContainer sampleData(byte[][] tmpImagesArr, int[] tmpLabelsArr, int totalExamples, List<Integer> iteratorDistribution, int maxTestSamples, int seed) {
        final byte[][] imagesArr = new byte[totalExamples][tmpImagesArr[0].length];
        final int[] labelsArr = new int[totalExamples];
        int count = 0;
        for (int label = 0; label < iteratorDistribution.size(); label++) {
            final int maxSamplesOfLabel = iteratorDistribution.get(label);
            /*
            Absolute HORRIBLE code!! Seriously considering converting this java code to Kotlin
            Java does not support Math.min of 3 integers
            Java does not provide a simple function to shuffle an array
            Java does not provide streams for raw bytes
            Java does not provide a simple function to convert an int array to a list or the other way around
            ...
             */
            final int[] matchingImageIndices = findMatchingImageIndices(label, tmpLabelsArr);
            final int[] shuffledMatchingImageIndices = shuffle(matchingImageIndices, seed);
            for (int j = 0; j < min(shuffledMatchingImageIndices.length, maxSamplesOfLabel, maxTestSamples); j++) {
                imagesArr[count] = tmpImagesArr[shuffledMatchingImageIndices[j]];
                labelsArr[count] = tmpLabelsArr[shuffledMatchingImageIndices[j]];
                count++;
            }
        }
        return new ImageLabelContainer(imagesArr, labelsArr);
    }

    public byte[] readImageUnsafe(int i) {
        return imagesArr[i];
    }

    public int readLabel(int i) {
        return labelsArr[i];
    }

    public int getNumSamples() {
        assert imagesArr.length == labelsArr.length;
        return imagesArr.length;
    }

    /**
     * Get the underlying images file as {@link MnistImageFile}.
     *
     * @return {@link MnistImageFile}.
     */
    public MnistImageFile getImages() {
        return mnistImageFile;
    }

    /* Why??? does Java not support tuples??? */
    private static class ImageLabelContainer {
        private byte[][] images;
        private int[] labels;

        public ImageLabelContainer(byte[][] images, int[] labels) {
            this.images = images;
            this.labels = labels;
        }
    }
}

File: fedml/src/main/java/nl/tudelft/trustchain/fedml/ai/dataset/mnist/CustomMnistManager.kt
package nl.tudelft.trustchain.fedml.ai.dataset.mnist

import mu.KotlinLogging
import nl.tudelft.trustchain.fedml.Behaviors
import nl.tudelft.trustchain.fedml.ai.dataset.DatasetManager
import org.deeplearning4j.datasets.mnist.MnistImageFile
import org.deeplearning4j.datasets.mnist.MnistLabelFile
import kotlin.random.Random

private val logger = KotlinLogging.logger("CustomMnistManager")

class CustomMnistManager(
    val imagesFile: String,
    val labelsFile: String,
    numExamples: Int,
    iteratorDistribution: IntArray,
    maxTestSamples: Int,
    seed: Long,
    behavior: Behaviors,
) : DatasetManager() {
    private val sampledImages: Array<FloatArray>
    private val sampledLabels: IntArray

    init {
        fullImagesArr.computeIfAbsent(imagesFile) { loadImages(imagesFile, numExamples) }
        fullLabelsArr.computeIfAbsent(labelsFile) { loadLabels(labelsFile, numExamples) }
        labelIndexMappings.computeIfAbsent(labelsFile) { generateLabelIndexMapping(fullLabelsArr[labelsFile]!!) }
        val imagesArr = fullImagesArr[imagesFile]!!
        val labelIndexMapping = labelIndexMappings[labelsFile]!!

        val labelIndexMapping2 = if (behavior === Behaviors.LABEL_FLIP) {
            val labelIndexMapping2 = labelIndexMapping.map { it.copyOf() }.toTypedArray()
            labelIndexMapping2[1] = labelIndexMapping[2]
            labelIndexMapping2[2] = labelIndexMapping[3]
            labelIndexMapping2[3] = labelIndexMapping[1]
            labelIndexMapping2
        } else {
            labelIndexMapping
        }
        val totalExamples = calculateTotalExamples(labelIndexMapping2, iteratorDistribution, maxTestSamples)
        val res = sampleData(imagesArr, totalExamples, iteratorDistribution, maxTestSamples, seed, labelIndexMapping2)
        sampledImages = res.first
        sampledLabels = res.second
    }

    private fun sampleData(
        tmpImagesArr: Array<ByteArray>,
        totalExamples: Int,
        iteratorDistribution: IntArray,
        maxTestSamples: Int,
        seed: Long,
        labelIndexMapping: Array<IntArray>,
    ): Pair<Array<FloatArray>, IntArray> {
        val placeholder = Pair(byteArrayOf(), -1)
        val results = Array(totalExamples) { placeholder }
        var count = 0
        val random = Random(seed)
        for (label in iteratorDistribution.indices) {
            val maxSamplesOfLabel = iteratorDistribution[label]
            labelIndexMapping[label].shuffle(random)
            for (j in 0 until minOf(labelIndexMapping[label].size, maxSamplesOfLabel, maxTestSamples)) {
                results[count] = Pair(tmpImagesArr[labelIndexMapping[label][j]], label)
                count++
            }
        }
        results.shuffle(random)
        val imagesArr = Array(totalExamples) { byteArrayOf() }
        val labelsArr = IntArray(totalExamples)
        results.forEachIndexed { i, pair ->
            imagesArr[i] = pair.first
            labelsArr[i] = pair.second
        }
        val imagesArr2 = imagesArr.map {
            it.map { byte ->
                (byte.toInt() and 0xFF).toFloat()
            }.toFloatArray()
        }.toTypedArray()
        return Pair(imagesArr2, labelsArr)
    }

    fun createTestBatches(): Array<Array<FloatArray>> {
        return (0 until 10).map { label ->
            val correspondingImageIndices = sampledLabels.indices
                .filter { i: Int -> sampledLabels[i] == label }
                .take(CustomMnistDataFetcher.TEST_BATCH_SIZE)
                .toTypedArray()
            correspondingImageIndices
                .map { i -> sampledImages[i] }
                .toTypedArray()
        }.toTypedArray()
    }

    fun readEntry(i: Int): Pair<FloatArray, Int> {
        return Pair(sampledImages[i], sampledLabels[i])
    }

    fun getNumSamples(): Int {
        return sampledImages.size
    }

    fun getLabels(): List<String> {
        return sampledLabels
            .distinct()
            .map { i: Int -> i.toString() }
    }

    fun getInputColumns(): Int {
        return getImageEntryLength(imagesFile)
    }

    companion object {
        private val imageEntryLength = hashMapOf<String, Int>()
        private val fullImagesArr = mutableMapOf<String, Array<ByteArray>>()
        private var fullLabelsArr = mutableMapOf<String, IntArray>()
        private var labelIndexMappings = mutableMapOf<String, Array<IntArray>>()

        @Synchronized
        private fun loadImages(filename: String, numExamples: Int): Array<ByteArray> {
            val file = MnistImageFile(filename, "r")
            imageEntryLength[filename] = file.entryLength
            return file.readImagesUnsafe(numExamples)
        }

        @Synchronized
        private fun loadLabels(filename: String, numExamples: Int): IntArray {
            val file = MnistLabelFile(filename, "r")
            return file.readLabels(numExamples)
        }

        @Synchronized
        private fun generateLabelIndexMapping(labelsArr: IntArray): Array<IntArray> {
            return (0 until CustomMnistDataFetcher.NUM_LABELS).map { label ->
                labelsArr.indices.filter { labelsArr[it] == label }.toIntArray()
            }.toTypedArray()
        }

        @Synchronized
        private fun getImageEntryLength(imagesFile: String): Int {
            return imageEntryLength[imagesFile]!!
        }
    }
}
